analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/TestAnalysisComparer.cs

```diff
@@ -0,0 +1,132 @@
+
+namespace SonarAnalyzer.UnitTest.TestFramework
+{
+    public static class TestAnalysisComparer
+    {
+        public static StringBuilder Compare(Diagnostic[] diagnostics, Dictionary<string, IList<IIssueLocation>> expectedIssuesPerFile, string languageVersion)
+        {
+            var summary = new StringBuilder($"Comparing actual issues with expected:\nLanguage version: {languageVersion}.\n");
+            var actualIssues = diagnostics.ToIssueLocations();
+
+            foreach (var fileName in actualIssues.Keys.OrderBy(x => x))
+            {
+                summary.Append('\n').Append(fileName).Append('\n');
+
+                foreach (var actualIssue in actualIssues[fileName].OrderBy(x => x.LineNumber))
+                {
+                    var issue = expectedIssuesPerFile[fileName].FirstOrDefault(x => x.IsPrimary == actualIssue.IsPrimary && x.LineNumber == actualIssue.LineNumber);
+                    summary.Append(Compare(actualIssue, issue));
+                    expectedIssuesPerFile[fileName].Remove(issue);
+                }
+
+                foreach (var expectedIssue in expectedIssuesPerFile[fileName].OrderBy(x => x.LineNumber))
+                {
+                    summary.Append($"Line {expectedIssue.LineNumber}: expected {IssueType(expectedIssue.IsPrimary)} issue {expectedIssue.IssueId ?? "(no id)"} was not raised!\n");
+                }
+            }
+
+            return summary;
+        }
+
+        private static StringBuilder Compare(IIssueLocation actual, IIssueLocation expected)
+        {
+            var summary = new StringBuilder();
+            var prefix = $"Line {actual.LineNumber}";
+            var issueType = IssueType(actual.IsPrimary);
+            var issueId = expected?.IssueId ?? actual.IssueId ?? "(no id)";
+
+            if (expected == null)
+            {
+                summary.Append($"{prefix}: unexpected {issueType} issue {issueId} with message '{actual.Message}'.\n");
+            }
+            else if (expected.Message != null && actual.Message != expected.Message)
+            {
+                summary.Append($"{prefix}: {issueType} issue {issueId} message '{expected.Message}' does not match the actual message '{actual.Message}'.\n");
+            }
+            else if (expected.Start.HasValue && actual.Start != expected.Start)
+            {
+                summary.Append($"{prefix}: {issueType} issue {issueId} should start on column {expected.Start} but got column {actual.Start}.\n");
+            }
+            else if (expected.Length.HasValue && actual.Length != expected.Length)
+            {
+                summary.Append($"{prefix}: {issueType} issue {issueId} should have a length of {expected.Length} but got a length of {actual.Length}.\n");
+            }
+
+            return summary;
+        }
+
+        private static IIssueLocation ToIssueLocation(this Diagnostic diagnostic) =>
+            new IssueLocationCollector.IssueLocation
```

Comment:
Let's give `IssueCollection` two handy constructors that will simplify this syntax. 
```
new IssueLocation(diagnostics) : this(diagnostics.GetMessage, diagnostics.Location){
IsPrimary=true
IssueId = diagnostics.Id
}
```
that will invoke the second one that also suits the overload below `(message, location)`

Reply:
Actually, once you do this, we will not even need these noisy workarounds (methods) at all.

Reply:
This is soooo much better 😆